File: backend/schemas/relations.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, model_validator


RelationKind = Literal["contains", "equals", "disjoint", "overlaps_unknown"]


class RelationCreate(BaseModel):
    from_flow_id: int
    to_flow_id: int
    kind: RelationKind
    note: str | None = None

    @model_validator(mode="after")
    def _no_self(self):
        if self.from_flow_id == self.to_flow_id:
            raise ValueError("a flow cannot relate to itself")
        return self


class RelationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_flow_id: int
    from_label: str | None = None
    to_flow_id: int
    to_label: str | None = None
    kind: str
    note: str | None
    created_at: datetime


class RelationCandidate(BaseModel):
    """A suggested (not yet stored) relation the user can confirm one-click."""

    other_flow_id: int
    other_label: str
    other_count: int | None = None
    other_period: str | None = None
    # Ready-to-POST proposal:
    from_flow_id: int
    to_flow_id: int
    kind: RelationKind
    reason: str

File: frontend/src/api/relations.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "./client";

export type RelationKind = "contains" | "equals" | "disjoint" | "overlaps_unknown";

export interface Relation {
  id: number;
  from_flow_id: number;
  from_label: string | null;
  to_flow_id: number;
  to_label: string | null;
  kind: RelationKind;
  note: string | null;
  created_at: string;
}

export interface RelationCandidate {
  other_flow_id: number;
  other_label: string;
  other_count: number | null;
  other_period: string | null;
  from_flow_id: number;
  to_flow_id: number;
  kind: RelationKind;
  reason: string;
}

export interface RelationCreatePayload {
  from_flow_id: number;
  to_flow_id: number;
  kind: RelationKind;
  note?: string | null;
}

export function useFlowRelations(flowId: number | null) {
  return useQuery<Relation[]>({
    queryKey: ["flow-relations", flowId],
    enabled: flowId != null,
    queryFn: async () => (await api.get(`/flow-relations?flow_id=${flowId}`)).data,
  });
}

export function useRelationCandidates(flowId: number | null) {
  return useQuery<RelationCandidate[]>({
    queryKey: ["flow-relation-candidates", flowId],
    enabled: flowId != null,
    queryFn: async () => (await api.get(`/flow-relations/candidates?flow_id=${flowId}`)).data,
  });
}

export function useCreateRelation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: RelationCreatePayload): Promise<Relation> =>
      (await api.post("/flow-relations", payload)).data,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["flow-relations"] });
      qc.invalidateQueries({ queryKey: ["flow-relation-candidates"] });
    },
  });
}

export function useDeleteRelation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: number) => api.delete(`/flow-relations/${id}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["flow-relations"] });
      qc.invalidateQueries({ queryKey: ["flow-relation-candidates"] });
    },
  });
}
